blueocean-rest-impl/src/test/java/io/jenkins/blueocean/util/HttpRequest.java

```diff
@@ -0,0 +1,225 @@
+import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.Map;
+
+import static io.jenkins.blueocean.util.HttpRequest.Method.GET;
+
+/**
+ * TODO: needs javadoc once API is final
+ * @author cliffmeyers
+ */
+public class HttpRequest {
+
+    private Logger logger = Logger.getLogger(getClass());
+    private static String JSON = "application/json";
+
+    enum Method { GET, POST, PUT, PATCH, DELETE };
+
+    final private Map<String, String> headers = new HashMap<>();
+    final private Map<String, Object> urlParts = new HashMap<>();
+    private Method method = GET;
+    private int expectedStatus = 200;
+    private String baseUrl = "";
+    private String requestUrl;
+    private String contentType;
+    private String requestBody;
+    private String username;
+    private String password;
+
+
+    private HttpRequest() {}
+
+    public static HttpRequest build(String baseUrl) {
+        Preconditions.checkState(StringUtils.isNotBlank(baseUrl), "baseUrl is required");
+        HttpRequest builder = new HttpRequest();
+        builder.baseUrl(baseUrl);
+        return builder;
+    }
+
+    public HttpRequest baseUrl(@Nonnull String baseUrl) {
+        this.baseUrl = baseUrl;
+        return this;
+    }
+
+    public HttpRequest GET(String url) {
+        this.method = GET;
+        this.requestUrl = url;
+        return this;
+    }
+
+    public HttpRequest POST(String url) {
+        this.method = Method.POST;
+        this.requestUrl = url;
+        return this;
+    }
+
+    public HttpRequest PUT(String url) {
+        this.method = Method.PUT;
+        this.requestUrl = url;
+        return this;
+    }
+
+    public HttpRequest PATCH(String url) {
+        this.method = Method.PATCH;
+        this.requestUrl = url;
+        return this;
+    }
+
+    public HttpRequest DELETE(String url) {
+        this.method = Method.DELETE;
+        this.requestUrl = url;
+        return this;
+    }
+
+    public HttpRequest urlPart(String key, String value) {
+        urlParts.put(key, value);
+        return this;
+    }
+
+    public HttpRequest header(String name, String value) {
+        headers.put(name, value);
+        return this;
+    }
+
+    public HttpRequest auth(String username, String password) {
+        this.username = username;
+        this.password = password;
+        return this;
+    }
+
+    public HttpRequest contentType(String contentType) {
+        this.contentType = contentType;
+        return this;
+    }
+
+    public HttpRequest bodyJson(Object object) {
+        if (object != null) {
+            contentType(JSON);
+            requestBody = JsonConverter.toJson(object);
+        } else {
+            contentType(null);
+            requestBody = null;
+        }
+        return this;
+    }
+
+    public HttpRequest status(int statusCode) {
+        this.expectedStatus = statusCode;
+        return this;
+    }
+
+    public <T> T asObject(Class<T> clazz) throws IOException {
```

Comment:
Its using template class, so maybe call it just `as`?

Reply:
I think it's good to leave this name longer so it's more easily differentiated between `asText`, `asBytes` and `asInputStream` methods.

Reply:
Except its not Object. Its generic typed return where as others are specific type return like String or InputStream. Fact that it doesn't return Object so calling it asObject() is confusing.

Reply:
It returns an instance that inherits from Object, no? ;)  I see your point. Would `asType` be better? I just think a method name of `as` isn't great, but maybe it's fine since it always includes a type as an arg?